### ISSUE TITLE
Log messages do not give useful location

### ISSUE THREAD

**Issue by bboreham** [cortexproject/cortex#1230]
Example:
```
level=warn ts=2019-02-04T09:21:37.646512093Z caller=gokit.go:46 method=/cortex.Ingester/Query duration=216.571µs err=EOF msg="gRPC\n"
```

`gokit.46` is utility code; we want to know where it was called from.

I believe the code does some counting of stack frames to arrive at the `caller`, so we need to set things up in a different order or adjust the count. 

Note some other messages do work as expected:
```
level=info ts=2019-02-04T09:21:35.849706809Z caller=lifecycler.go:441 msg="changing ingester state from" old_state=JOINING new_state=ACTIVE
```

Very similar to #655

**Comment by bboreham** [comment]
After #1297 I get:

```
level=warn ts=2019-03-28T11:02:28.071677812Z caller=chain.go:39 method=/cortex.Ingester/Query duration=127.475902ms err="context canceled" msg="gRPC\n"
```

which doesn't seem right either.  That file is `vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go`; the real caller is `vendor/github.com/weaveworks/common/middleware/grpc_logging.go`

**Comment by yiyuanliu** [comment]
It seems that this issue is related to the version of go. Wrapper functions generated by compiler are no longer reported in go 1.12.0. That means we can ignore the `<autogenerated> function` and just use `log.Caller(4)` in go 1.12.0.

#1297 works in go 1.10 but not in go 1.12. Do we need to support old go versions? If not, we can just change from `log.Caller(5)` to `log.Caller(4)`.

**Comment by bboreham** [comment]
I think it would be ok to support 1.12 and add a comment noting it doesn’t work right in 1.10

Could use `//build` to say it doesn’t build on 1.10. Cortex as a project has moved to 1.12.

**Comment by bboreham** [comment]
I think this was fixed by #1297 and #1308
